### PR TITLE
Change kani-compiler to use `-C llvm-args` for Kani specific flags

### LINKED ISSUE #2137: Improve how we pass Kani codegen specific options to `kani-compiler`
**Proposed change:** Instead of using an environment variable or relying on the arguments order to pass information to `kani-compiler`, we should either use `-C llvm-args=` or (ideally) help create `-C backed-args=` and use them with `kani` backend (as [discussed here](https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/Tools.20.2F.20custom.20backend.20arguments)). 

**Motivation:** This is a blocker for reusing compilation artifacts in Kani (#2036). The current mechanism doesn't work well with Cargo's caching. Cargo uses the provided `rustc` flags to compute its artifacts cache, and to compute when it needs to recompile an artifact. This is one of the reasons we explicitly do a `cargo clean` every time users invoke `cargo kani`.

This is also hacky and rather fragile.

### PR DESCRIPTION
**Pull request by celinval**
### Description of changes: 

Changed how `kani-driver` passess arguments to `kani-compiler`. We now have one option, `--kani-compiler`, that changes how we configure rustc, and all the other Kani specific arguments are added to `-C llvm-args`. We no longer need to keep the arguments in a certain order or need to use KANI_FLAGS environment variable.

In the `kani-compiler` we no longer parse the arguments before invoking rustc, we initially detect whether users want to use the Kani flavour of the compiler by looking for a `--kani-compiler` flag.

If `--kani-compiler` flag is found, we configure rustc's driver to use our custom backend and our custom callback. We then invoke the driver with the command line arguments excluding the `--kani-compiler` flag. During the callback configuration we initialize the Kani specific flags that are passed via `-C llvm-args`.

If there isn't any `--kani-compiler` flag, we just invoke rustc with all the given arguments.

### Resolved issues:

Resolves #2137 

### Related RFC:

Optional #ISSUE-NUMBER.

### Call-outs:

Since we initialize the QueryDb after sharing it, I put the entire structure behind a mutex, and converted its members to simple non-synchronized values.

### Testing:

* How is this change tested? Existing tests

* Is this a refactor change? Yes

### Checklist
- [ ] Each commit message has a non-empty body, explaining why the change was made
- [ ] Methods or procedures are documented
- [ ] Regression or unit tests are included, or existing tests cover the modified code
- [ ] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: kani-compiler/src/codegen_cprover_gotoc/compiler_interface.rs [hunk 1]
```diff
@@ -45,19 +45,19 @@ use std::io::Write as IoWrite;
...
-use std::rc::Rc;
+use std::sync::{Arc, Mutex};
 use std::time::Instant;
 use tempfile::Builder as TempFileBuilder;
 use tracing::{debug, error, info};
 
 #[derive(Clone)]
 pub struct GotocCodegenBackend {
-    queries: Rc<QueryDb>,
+    queries: Arc<Mutex<QueryDb>>,
```
Comment (zhassan-aws) [inline kani-compiler/src/codegen_cprover_gotoc/compiler_interface.rs:55]: Can you add a comment that explains why this needs Arc/Mutex?

File: kani-compiler/src/main.rs [hunk 2]
```diff
@@ -31,152 +31,27 @@ extern crate tempfile;
...
-        assert_eq!(os_str.to_str(), None);
-
-        let matches = parser::parser().get_matches_from(vec![
-            OsString::from("kani-compiler").as_os_str(),
-            OsString::from("--sysroot").as_os_str(),
-            OsString::from("any").as_os_str(),
-            os_str,
-        ]);
-        generate_rustc_args(&matches);
+    if kani_compiler {
```
Comment (zhassan-aws) [inline kani-compiler/src/main.rs:50]: Should this be an assert instead? Are there cases where we invoke the driver without `--kani-compiler`?

File: kani-compiler/src/kani_compiler.rs [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    Box::new(GotocCodegenBackend::new(queries))
+}
+
+/// Fallback backend. It will trigger an error if no backend has been enabled.
+#[cfg(not(feature = "cprover"))]
+fn backend(queries: Arc<Mutex<QueryDb>>) -> Box<CodegenBackend> {
+    compile_error!("No backend is available. Only supported value today is `cprover`");
+}
+
+/// Empty struct since we don't support any callbacks yet.
```
Comment (zhassan-aws) [inline kani-compiler/src/kani_compiler.rs:65]: The comment is misplaced/outdated.

File: kani-driver/src/call_single_file.rs [hunk 4]
```diff
@@ -160,6 +166,17 @@ impl KaniSession {
...
+        // rustc driver is invoked.
+        flags.push("--kani-compiler".into());
+
         flags
     }
 }
+
+/// This function can be used to convert a Kani compiler specific argument into a rustc one.
+/// We currently pass Kani specific arguments using the `--llvm-args` structure which is the
+/// hacky mechanism used by other rustc backend to receive arguments unknown to rustc.
```
Comment (zhassan-aws) [inline kani-driver/src/call_single_file.rs:179]: Can you elaborate on how llvm-args affects the cache?

**Review by zhassan-aws — COMMENTED** [review]
LGTM. A few minor comments.

File: kani-compiler/src/main.rs (hunk 2 above)
Comment (celinval) [reply kani-compiler/src/main.rs:50]: Yes. To build proc_macro dependencies and build scripts. Cargo also invokes without ``--kani-compiler` to retrieve information about the compiler such as version.

File: kani-driver/src/call_single_file.rs (hunk 4 above)
Comment (celinval) [reply kani-driver/src/call_single_file.rs:179]: Sure

**Comment by celinval** [comment]
I updated the comments. Let me know if that makes sense.

File: kani-driver/src/call_single_file.rs [hunk 5]
```diff
@@ -174,9 +174,16 @@ impl KaniSession {
     }
 }
 
-/// This function can be used to convert a Kani compiler specific argument into a rustc one.
+/// This function can be used to convert Kani compiler specific arguments into a rustc one.
 /// We currently pass Kani specific arguments using the `--llvm-args` structure which is the
 /// hacky mechanism used by other rustc backend to receive arguments unknown to rustc.
+///
+/// Note that Cargo caching mechanism takes the building context into consideration, which
```
Comment (zhassan-aws) [inline kani-driver/src/call_single_file.rs:181]: Thanks!

**Review by zhassan-aws — APPROVED** [review]

**Pull request merged**